homeassistant/components/sensor/enphase_envoy.py

```diff
@@ -0,0 +1,133 @@
+"""
+Support for monitoring energy usage and solar panel energy production using
+the Enphase Envoy.
+
+For more details about this platform, please refer to the documentation at
+
+"""
+import logging
+import json
+
+import voluptuous as vol
+
+from homeassistant.helpers.entity import Entity
+from homeassistant.components.sensor import PLATFORM_SCHEMA
+import homeassistant.helpers.config_validation as cv
+
+
+_LOGGER = logging.getLogger(__name__)
+
+CONF_IP_ADDRESS = 'ip'
+CONF_MONITORED_CONDITIONS = 'monitored_conditions'
+
+DEFAULT_NAMES = [
+    "Envoy Current Energy Production",
+    "Envoy Today's Energy Production",
+    "Envoy Last Seven Days Energy Production",
+    "Envoy Lifetime Energy Production",
+    "Envoy Current Energy Consumption",
+    "Envoy Today's Energy Consumption",
+    "Envoy Last Seven Days Energy Consumption",
+    "Envoy Lifetime Energy Consumption"]
+
+SENSOR_TYPES = [
+    "production", "daily_production", "7_days_production",
+    "lifetime_production", "consumption", "daily_consumption",
+    "7_days_consumption", "lifetime_consumption"]
+
+ICON = 'mdi:flash'
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Required(CONF_IP_ADDRESS): cv.string,
+    vol.Optional(CONF_MONITORED_CONDITIONS): vol.All(cv.ensure_list)
+
+})
+
+
+def setup_platform(hass, config, add_devices, discovery_info=None):
+    """Set up the Enphase Envoy sensor."""
+    ip_address = config.get(CONF_IP_ADDRESS)
+    monitored_conditions = config.get(CONF_MONITORED_CONDITIONS, {})
+
+
+# Iterate through the list of sensors, adding it if either monitored
+# conditions has been left out of the config, or that the given sensor
+# is in the monitored conditions list
+    for i in range(8):
+        if monitored_conditions == {} or \
+            SENSOR_TYPES[i] in monitored_conditions:
+                add_devices([Envoy(ip_address, DEFAULT_NAMES[i],
+                            SENSOR_TYPES[i])], True)
+
+
+class Envoy(Entity):
+    """Implementation of the Enphase Envoy sensors."""
+    def __init__(self, ip_address, name, type):
+        """Initialize the sensor."""
+        self._url = "http://{}/production.json".format(ip_address)
+        self._name = name
+
+        if type == 'production' or type == 'consumption':
+            self._unit_of_measurement = 'watts'
+        else:
+            self._unit_of_measurement = "watt hours"
```

Comment:
I would prefer `W` and `Wh` here. Looking at `homeassistant/const.py` it seems to be the consensus for other units of measurement to use the abbreviated version. Maybe these two could be added there?

Reply:
A couple of really good points, thank you. I'm not really sure how to do unit tests. Is there somewhere you can suggest to learn more about that?

Reply:
Maybe the easiest way is by looking at the existing unit tests. Most unit tests focus on two areas - setting up the component, and then testing the various functionalities, transformation, error handling, etc. You can load a static JSON file with `load_fixture`.
The [developer documentation tells you how to run tests](https://developers.home-assistant.io/docs/en/development_testing.html) in your environment.